### ISSUE TITLE
[Bug] reset_state() fails to clear accumulated confusion matrix in MeanIoU

### ISSUE THREAD

**Issue by amadhan882** [keras-team/keras#22836]
## Description
The `reset_state()` method in `MeanIoU` does not properly zero out the underlying state variables. This causes metric accumulation to persist across calls where a fresh state is expected, such as between different evaluation runs or epochs.

## Reproduction Script

```python
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras
import numpy as np

metric = keras.metrics.MeanIoU(num_classes=3)


y_true1 = np.array([0, 1, 2])
y_pred1 = np.array([0, 1, 2])
metric.update_state(y_true1, y_pred1)
result1 = metric.result().numpy()
print(f"Result Batch 1 (Expected 1.0): {result1}")

metric.reset_state()

y_true2 = np.array([1, 2, 0])
y_pred2 = np.array([1, 2, 0])
metric.update_state(y_true2, y_pred2)
result2 = metric.result().numpy()
print(f"Result Batch 2 after reset (Expected 1.0): {result2}")


if np.allclose(result1, result2):
    print(" Reset_state() failed to clear the internal confusion matrix. ")
```

## Observed behavior
The metric continues to accumulate values from previous states even after `reset_state()` is called. In some backend configurations, the result of Batch 2 is identical to an accumulation of Batch 1 + Batch 2, rather than a fresh calculation.

## Expected behavior
Calling `reset_state()` should completely clear the internal confusion matrix (setting all elements to zero). Subsequent calls to `result()` after a new `update_state()` should only reflect the data provided after the reset.

## Notes
This issue is particularly dangerous for long-running training jobs where researchers rely on epoch-by-epoch metric accuracy. It may be related to how the backend variables are handled during the `assign` operations in the reset logic.

**Comment by maitry63** [comment]
Hi @amadhan882 , Thank you for providing the reproduction script. I checked the metric state in detail before and after `reset_state()`.

Check this [gist ](https://colab.sandbox.google.com/gist/maitry63/1ba0b870a01942f37b43b01add353cf4/untitled5.ipynb#scrollTo=Rt6OB-Lnr714)for details.

It seems, 
- `reset_state()` works correctly
- The internal confusion matrix (total_cm) is fully cleared after ` reset_state()`.
- No state from previous batches carries over.
- Identical results across independent perfect batches are expected, as MeanIoU depends on per-class overlap.
- Small differences from 1.0 (0.9999998) are due to floating-point precision and are not errors.

The behavior in the original script is expected. `MeanIoU.reset_state()` clears the metric correctly, and subsequent updates are computed independently. The perceived “reset failure” comes from misunderstanding metric invariance and floating-point rounding, not from a bug.

**Comment by amadhan882** [comment]
Hi @maitry63,

Thanks for the feedback. I see the confusion—using two "perfect" batches in the original script made it impossible to distinguish between a fresh 1.0 and a cumulative 1.0.

The issue isn't the floating-point precision, but rather the **state retention**. To prove this, we need a **differential test** where Batch 2 is a total mismatch. If the reset works, the second result must be 0.0. If it fails, the "perfect" history from Batch 1 will buffer the result.

```python
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras
import numpy as np

metric = keras.metrics.MeanIoU(num_classes=2)

# 1. Batch 1: Perfect Match (Should be 1.0)
metric.update_state([0, 1], [0, 1])
print(f"Batch 1 result: {metric.result().numpy()}") 

# 2. Reset State
metric.reset_state()

# 3. Batch 2: Complete Mismatch (Should be 0.0)
# If reset failed, the result will be ~0.33 or ~0.5 instead of 0.0
metric.update_state([0, 1], [1, 0]) 
print(f"Batch 2 result (post-reset): {metric.result().numpy()}")

```
**The Point:**
If Batch 2 returns anything significantly higher than 0.0, it confirms the confusion matrix was not zeroed out. This often happens in specific backend/distributed configurations where variable assignments don't sync properly.
Could you run this asymmetric test and check if Batch 2 actually hits 0.0?

**Comment by maitry63** [comment]
Thanks for suggesting the asymmetric test ,  this helps clarify the behavior.

I  ran the test in this  [gist](https://colab.sandbox.google.com/gist/maitry63/1ba0b870a01942f37b43b01add353cf4/untitled5.ipynb#scrollTo=LMuwncpCQCgz) with a perfect batch followed by a reset and then a complete mismatch. The result after reset is exactly 0.0, which confirms that `reset_state()` is correctly clearing the internal confusion matrix.

If any state were retained, the second result would have been higher ,due to accumulated history.
This demonstrates that the metric is recomputed from a clean state after reset, and no accumulation occurs.

So we can say that the observed behavior is correct, and `MeanIoU.reset_state()` is functioning  correctly.

If this solves your doubt, please feel free to close this issue.
